syntax = "proto3";

package edu.uci.ics.amber.engine.common;

// sealed trait ActorVirtualIdentity
message ActorVirtualIdentity {
  oneof sealed_value {
    WorkerActorVirtualIdentity workerActorVirtualIdentity = 1;
    ControllerVirtualIdentity controllerVirtualIdentity = 2;
    SelfVirtualIdentity selfVirtualIdentity = 3;
    ClientVirtualIdentity clientVirtualIdentity = 4;
  }
}

// final case class WorkerActorVirtualIdentity (
//    name: String
// ) extends ActorVirtualIdentity
message WorkerActorVirtualIdentity {
  string name = 1;
}

// final case class ControllerVirtualIdentity extends ActorVirtualIdentity
message ControllerVirtualIdentity {}

// final case class SelfVirtualIdentity extends ActorVirtualIdentity
message SelfVirtualIdentity {}

// final case class ClientVirtualIdentity extends ActorVirtualIdentity
message ClientVirtualIdentity {}

// final case class LayerIdentity (
//    workflow: String,
//    operator: String,
//    layerID: String
// )
message LayerIdentity{
  string workflow = 1;
  string operator = 2;
  string layerID = 3;
}

// final case class LinkIdentity (
//    from: LayerIdentity,
//    to: LayerIdentity
// )
message LinkIdentity {
  LayerIdentity from = 1 ;
  LayerIdentity to = 2;
}

// final case class LinkIdentity (
//    workflow: String,
//    operator: String
// )
message OperatorIdentity {
  string workflow = 1;
  string operator = 2;
}

// final case class WorkflowIdentity (
//    id: String
// )
message WorkflowIdentity {
  string id = 1;
}

